### PR TITLE
Update sampling code, add exponential (biased) sampling

### PR DESCRIPTION
**Pull request by Bounciness**
This commit reorganized the sampling code into a submodule in external/Volume-and-Sampling/, which is maintained here: https://github.com/Bounciness/Volume-and-Sampling. Before this merge, I was separately maintaining two versions of the sampling code (one for COBRA, one for general use). This update merges the two functionalities. If the COBRA wrappers are used, everything should behave identically to before.

The main new feature is that exponential (biased) sampling of the solution space is supported. This update includes a few minor bug fixes with the sampling code. 

**I hereby confirm that I have:**

- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [X] Selected `develop` as a target branch (top left drop-down menu)

*(Note: You may replace [] with [X] to check the box)*

### CODE REVIEW

File: src/analysis/sampling/CHRR/chrrSampler.m [hunk 1]
```diff
@@ -1,45 +1,40 @@
...
 % OUTPUTS:
-%    samples:             `n x numSamples` matrix of random flux samples
-%    roundedPolytope:     The rounded polytope. Save for use in subsequent
-%                         rounds of sampling.
-%    minFlux, maxFlux:    flux minima and maxima
-%
-% .. Authors: - Ben Cousins and Hulda S. Haraldsdóttir, March 2017
+% samples ... n x numSamples matrix of random flux samples
+% roundedPolytope ... The rounded polytope. Save for use in subsequent
+%                     rounds of sampling.
```
Comment (tpfau) [inline src/analysis/sampling/CHRR/chrrSampler.m]: Please keep to the defined Documentation scheme detailed in [this](https://github.com/opencobra/cobratoolbox/blob/master/docs/source/guides/documentationGuide.md) guide

File: src/analysis/sampling/CHRR/chrrSampler.m [hunk 2]
```diff
@@ -1,45 +1,40 @@
-function [samples, roundedPolytope, minFlux, maxFlux] = chrrSampler(model, numSkip, numSamples, toRound, roundedPolytope, minFlux, maxFlux, useFastFVA)
-% Generate uniform random flux samples with CHRR
-% Coordinate Hit-and-Run with Rounding
+function [samples,roundedPolytope] = chrrSampler(model,numSkip,numSamples,toRound,roundedPolytope,minFlux,maxFlux)
```
Comment (tpfau) [inline src/analysis/sampling/CHRR/chrrSampler.m]: If it is necessary to change the header (which is strongly discouraged), please make sure, that any tutorials and tests are updated accordingly. 
From what I see in the code, this change essentially removed the information in `P.A_eq` and `P.b_eq`. Is that intended? As outputs might change by this alteration.

File: src/analysis/sampling/CHRR/chrrParseModel.m [hunk 3]
```diff
@@ -1,34 +1,29 @@
...
+% P ... A structure with fields:
+% .A_eq ... Equality constraint matrix (model.S)
+% .b_eq ... Right hand side of equality constraints (model.b)
+% .A ... Inequality constraint matrix ([I_n 0; 0 -I_n])
+% .b ... Right hand side of inequality constraints ([lb; -ub])
+% 
+% October 2016, Ben Cousins and Hulda S. Haraldsdóttir
+%
+% --Moved objective function handling to preprocess function: Ben Cousins,
+% November 2017
```
Comment (tpfau) [inline src/analysis/sampling/CHRR/chrrParseModel.m]: Please format the documentation according to the [documentation guide](https://github.com/opencobra/cobratoolbox/blob/master/docs/source/guides/documentationGuide.md)

File: src/analysis/sampling/CHRR/chrrExpSampler.m [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+%                     sampling the same model.
+% minFlux,maxFlux ... n x 1 flux minima and maxima from flux variability
+%                     analysis of the same model.
+%
+% OUTPUTS:
+% samples ... n x numSamples matrix of random flux samples
+% roundedPolytope ... The rounded polytope. Save for use in subsequent
+%                     rounds of sampling.
+%
+% November 2017, Ben Cousins and Hulda S. Haraldsdóttir
```
Comment (tpfau) [inline src/analysis/sampling/CHRR/chrrExpSampler.m]: please format the documentation according to the [documentation guide](https://github.com/opencobra/cobratoolbox/blob/master/docs/source/guides/documentationGuide.md)

File: src/analysis/sampling/CHRR/chrrSampler.m [hunk 5]
```diff
@@ -78,33 +69,30 @@
...
-    
-    eps_cutoff = 1e-7;
-    
-    isEq = (maxFlux - minFlux) < eps_cutoff;
-    eq_constraints = sparse(sum(isEq),size(P.A_eq,2));
-    eq_constraints(:,isEq) = speye(sum(isEq));
-    
-    P.A_eq = [P.A_eq; eq_constraints];
-    P.b_eq = [P.b_eq; minFlux(isEq)];
+%     if toGetWidths
```
Comment (tpfau) [inline src/analysis/sampling/CHRR/chrrSampler.m]: Please remove large commented sections.

**Comment by tpfau** [comment]
Oh, also, could you add a test for the exponential sampling method?

**Comment by laurentheirendt** [comment]
@Bounciness ?

File: src/analysis/sampling/CHRR/chrrExpSampler.m (hunk 4 above)
Comment (Bounciness) [reply src/analysis/sampling/CHRR/chrrExpSampler.m]: This should be fixed now.

File: src/analysis/sampling/CHRR/chrrSampler.m (hunk 5 above)
Comment (Bounciness) [reply src/analysis/sampling/CHRR/chrrSampler.m]: Removed.

**Comment by Bounciness** [comment]
Thanks a lot for your feedback Thomas, sorry for delay getting back to you. I've added a test for exponential sampling in analysis/testSampling/testSampleCbModel.m. 

With regards to the function header being different, the restructuring of the code makes the previous function header unnecessarily complex (i.e. we can get rid of two arguments). I ran the sampling tutorial line-by-line, everything seems to be good. The only real change required was changing the function call to chrrSampler() in sampling/sampleCbModel.m. I also added a function call to chrrExpSampler() in sampleCbModel.m. If a user is doing sampling through sampleCbModel(), then everything should be equivalent to before.

Hopefully all the documentation guidelines are adhered to, let me know if there's anything I missed.

**Comment by tpfau** [comment]
Great, 
Would it be posible to add ethe exponential sampler as an option to sampleCbModel?
And could you update the tutorial?

**Comment by rmtfleming** [comment]
@Gpreciat is responsible for maintaining the tutorial, which can be updated once we have fully tested this new sampler functionality.

**Comment by Bounciness** [comment]
I've added CHRR_EXP as an option to sampleCbModel.

**Comment by laurentheirendt** [comment]
@Bounciness, may you solve the conflict here? Thanks 👍

**Comment by Bounciness** [comment]
@laurentheirendt I seem to be stuck in an infinite loop of conflicts. I thought I resolved it through the web interface on Github but it still says conflicting. Any suggestions?

**Comment by laurentheirendt** [comment]
@Bounciness, I worked on this PR and I submitted [this PR](https://github.com/Bounciness/Volume-and-Sampling/pull/1) to make it work.

Once this is in, I will continue working on this PR. Thanks 👍

**Comment by Bounciness** [comment]
@laurentheirendt, I'm terribly sorry! I forgot to commit my updates to preprocess.m on my submodule Volume-and-Sampling, which is why you were getting the errors. 

I've pushed my updated version to my submodule (which is essentially the version you submitted in your PR, but before calling fastFVA, it checks for CPLEX--this is the existing behavior in COBRA for sampling).

**Comment by laurentheirendt** [comment]
@Bounciness, no worries. I will close this PR and resubmit the updated one. Thanks 👍

**Pull request closed without merging**